ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteMergeRewriter.java

```diff
@@ -149,8 +153,20 @@ public void appendWhenNotMatchedInsertClause(MergeStatement.InsertClause insertC
         hintStr = null;
       }
       List<String> values = sqlGenerator.getDeleteValues(Context.Operation.MERGE);
-      values.add(insertClause.getValuesClause());
       
+      if (insertClause.getColumnListText() != null) {
+        String[] columnNames = insertClause.getColumnListText()
+            .substring(1, insertClause.getColumnListText().length() - 1).split(",");
+        String[] columnValues = insertClause.getValuesClause().split(",");
```

Comment:
Initially values are also represented in the AST
```
TOK_FUNCTION
  struct
  .
    TOK_TABLE_OR_COL
      some_alias
    a
  .
    TOK_TABLE_OR_COL
      some_alias
    c,c
```
We can try transform it to a `List<String>`